File: src/client/java/fr/hardel/asset_editor/client/javafx/components/page/enchantment/ExclusiveSingleSection.java
package fr.hardel.asset_editor.client.javafx.components.page.enchantment;

import fr.hardel.asset_editor.client.javafx.lib.FxSelectionBindings;
import fr.hardel.asset_editor.client.javafx.lib.StudioContext;
import fr.hardel.asset_editor.client.selector.StoreSelection;
import fr.hardel.asset_editor.network.EditorAction;
import fr.hardel.asset_editor.store.ElementEntry;
import javafx.scene.layout.VBox;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.List;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.UnaryOperator;

public final class ExclusiveSingleSection extends VBox {

    public ExclusiveSingleSection(StudioContext context,
        StoreSelection<ElementEntry<?>, Set<String>> directExclusiveSelection,
        FxSelectionBindings bindings,
        BiFunction<UnaryOperator<Enchantment>, EditorAction, Boolean> applyMutation) {
        setSpacing(32);
        setMaxWidth(Double.MAX_VALUE);

        List<Identifier> custom = context.allTypedEntries(Registries.ENCHANTMENT).stream()
            .map(ElementEntry::id)
            .filter(id -> !id.getNamespace().equals("minecraft"))
            .toList();

        List<Identifier> vanilla = context.allTypedEntries(Registries.ENCHANTMENT).stream()
            .map(ElementEntry::id)
            .filter(id -> id.getNamespace().equals("minecraft"))
            .toList();

        if (!vanilla.isEmpty()) {
            getChildren().add(new EnchantmentCategory(
                I18n.get("enchantment.exclusive:vanilla"),
                vanilla,
                directExclusiveSelection,
                bindings,
                applyMutation,
                context));
        }
        getChildren().add(new EnchantmentCategory(
            I18n.get("enchantment.exclusive:custom"),
            custom,
            directExclusiveSelection,
            bindings,
            applyMutation,
            context));
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/page/enchantment/ExclusiveSingleSection.kt
package fr.hardel.asset_editor.client.compose.components.page.enchantment

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import fr.hardel.asset_editor.client.compose.lib.StudioContext
import fr.hardel.asset_editor.client.compose.lib.rememberRegistryEntries
import fr.hardel.asset_editor.client.memory.session.server.ClientWorkspaceRegistries
import net.minecraft.client.resources.language.I18n
import net.minecraft.resources.Identifier
import fr.hardel.asset_editor.AssetEditor

@Composable
fun ExclusiveSingleSection(
    context: StudioContext,
    directExclusiveIds: Set<String>,
    onToggleExclusive: (Identifier) -> Unit
) {
    val entries = rememberRegistryEntries(context, ClientWorkspaceRegistries.ENCHANTMENT)
    val allIds = remember(entries) { entries.map { entry -> entry.id() } }
    val custom = remember(allIds) { allIds.filter { id -> id.namespace != Identifier.DEFAULT_NAMESPACE } }
    val vanilla = remember(allIds) { allIds.filter { id -> id.namespace == Identifier.DEFAULT_NAMESPACE } }

    if (vanilla.isNotEmpty()) {
        EnchantmentCategory(
            title = I18n.get("enchantment.exclusive:vanilla"),
            identifiers = vanilla,
            directExclusiveIds = directExclusiveIds,
            onToggleExclusive = onToggleExclusive,
            context = context
        )
    }

    EnchantmentCategory(
        title = I18n.get("enchantment.exclusive:custom"),
        identifiers = custom,
        directExclusiveIds = directExclusiveIds,
        onToggleExclusive = onToggleExclusive,
        context = context
    )
}
